### ISSUE TITLE
live ping metrics

### ISSUE THREAD

**Issue by jakesullyneytiri** [CachetHQ/Cachet#1488]
Hi,

i know this site https://status.discordapp.com/ doesen't use cachet but it uses status.io or something like that and i was wondering is it possible to get a live ping stats thingy as that site but on cachet? so people can see actual ping ms and earlier stats??

**Comment by GrahamCampbell** [comment]
That would get very slow. Load times once we have more than about 10,000 points in the db are ~5 seconds.

**Comment by GrahamCampbell** [comment]
We're working on improving that first.

**Comment by jakesullyneytiri** [comment]
okay well it doesen't have to log the ping response more than to a month maximum so when new ping stats comes it will slowly earse oldest ping log from metrics and so on :)

**Comment by GrahamCampbell** [comment]
Yeh, we don't do that yet though. ;)

**Comment by GrahamCampbell** [comment]
And, even one ping stat per min, for 30 days would cause cachet to hang for like 30 seconds on load.

**Comment by GrahamCampbell** [comment]
That's ~45 thousand metric points.

**Comment by GrahamCampbell** [comment]
See https://github.com/CachetHQ/Cachet/issues/1455 regarding cleaning up old points.

**Comment by jakesullyneytiri** [comment]
well stats.io has it without any slow downs or issues with their site so i stil ldont think this is impossible to fix :)

**Comment by GrahamCampbell** [comment]
Yeh, it's totally doable. I'm just saying the current state of our metric isn't where we want them yet.

**Comment by lawrencegs** [comment]
If its just ping, how about integrates with provider that does ping for a living, e.g statuscake? They have API even for its free tier...

**Comment by GrahamCampbell** [comment]
@lawrencegs I'm pretty sure that wasn't the question. Of course Cachet won't be doing the pinging part?

**Comment by jakesullyneytiri** [comment]
well actually i want script to do pinging part so it can sort off keep metric's up to date but it doesen't have to ping ever 0 seconds it can do like every 5-50 seconds or so why not just add a setting in cachet that will let user with script be able to adjust how pinging shal be done like shall it do in 5 seconds or 15 seconds or so on.

**Comment by GrahamCampbell** [comment]
When you say "pinging", you don't mean "pinging", right? You just want to poll for changes?

**Comment by jakesullyneytiri** [comment]
the pinging is to fetch MS response so lets say first check was at 15:00 ms is 90ms then after w.e setting you told metrics to update then at 15:30 it has ms at 120ms and so on that way it can keep giving live ms response.

**Comment by makkesk8** [comment]
Instead of storing the metrics your self why not let users integrate with services that already does this? statuscake, nixstats(free), pingdom, uptimerobot(free) and just pull in response times? I don't see why you should even bother with live response time since it doesn't really matter.

**Comment by GrahamCampbell** [comment]
> Instead of storing the metrics your self why not let users integrate with services that already does this? statuscake, nixstats(free), pingdom, uptimerobot(free) and just pull in response times? I don't see why you should even bother with live response time since it doesn't really matter.

We already can integrate with those services.

**Comment by makkesk8** [comment]
Oh great, haven't used it yet so I did not know :P

**Comment by zhil** [comment]
@GrahamCampbell @jbrooksuk I have an idea regarding how we could do live monitoring for page generation speed on production. With very small outage.

1) get visitors stats for site from google-analytics api. Hits count is green, store average numbers per 5 minutes or so. (more simple and dirty solution - send UDP hits on all pageloads)
2) add UDP proxy transfer to client site. Each time when page generation time is bigger as first threshhold (0.5 seconds or so) client send UDP packet to proxy, proxy send TCP API request to CachetHQ. near to zero additional time for clients script generation.
3) when generation time hit second treshhold - increase third metric.

Database overhead - 3 metrics, 5 minutes averages - thats just 864 entities daily. This info is useless after 2-3 days, so it would be around 3k entities.
Some less detailed stats could be generated for longer stats.

It could look something like
![image](https://cloud.githubusercontent.com/assets/981783/13156766/c750ad2c-d68c-11e5-8d0b-682f4d201c59.png)

**Comment by zhil** [comment]
@GrahamCampbell @jbrooksuk
I have checked database and have another suggestion. For metrics with "average" stats you could store "treshhold" interval (like 5 minutes), value and counter.
So metrics would be optimized like
![image](https://cloud.githubusercontent.com/assets/981783/13157221/53c1db94-d68f-11e5-99e9-33359a5391f6.png)

When you will get new metrics point for metric with "average" stats and if there are another metric within treshhold - you could simply update stats instead of saving new point. On sample I got hit with value 0.8, so average would be

(0.58*9 + 0.8)/10=0,602

![image](https://cloud.githubusercontent.com/assets/981783/13157284/a6f8ba30-d68f-11e5-8f88-9bfa86abe71b.png)

**Comment by zhil** [comment]
it will allow to show metrics with averages for A LOT of points (like page average generation speed)

**Comment by jbrooksuk** [comment]
We'd accept a PR for your suggestion @zhil. There could be more overhead on it though, having to find and update the metric each time.

Definitely something to look at it.

**Comment by jbrooksuk** [comment]
Once #1582 is merged in, this is something we'll be able to actually think about :+1:

**Comment by jbrooksuk** [comment]
Could I get some feedback on #1582 please?

**Comment by jbrooksuk** [comment]
The idea I get from this is to recreate Pingdom. Cachet won't be monitoring any websites. But from this we're refactoring metric points, so thanks!

**Comment by irfanops** [comment]
@GrahamCampbell @jbrooksuk 
Still seeing lot of delay when loading ping metrics. Using cachet-monitor to update the ping metric on Cachet dashboard. Is this good practise? Please advise. Thanks!

**Comment by jbrooksuk** [comment]
@irfanops could you open as a new issue please. You'll need to tell us more about your setup too, how many Metrics (+ Points), Apache/Nginx, DB driver etc.
